server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -219,14 +258,24 @@ remoteSchemaUnion
   . (MonadSchema n m, MonadError QErr m)
   => SchemaIntrospection
   -> G.UnionTypeDefinition
-  -> m (Parser 'Output n ())
+  -> m (Parser 'Output n (SelectionSet NoFragments G.Name))
 remoteSchemaUnion schemaDoc defn@(G.UnionTypeDefinition description name _directives objectNames) =
   P.memoizeOn 'remoteSchemaObject defn do
-  objDefs <- traverse getObject objectNames
-  objs :: [Parser 'Output n ()] <- traverse (remoteSchemaObject schemaDoc) objDefs
+  objs :: [Parser 'Output n (Name, (SelectionSet NoFragments G.Name))] <-
+    traverse (\objName -> do
+                 obj <- getObject objName
+                 rsObj <- remoteSchemaObject schemaDoc obj
+                 return $ (objName,) <$> rsObj)
+             objectNames
   when (null objs) $
     throw400 RemoteSchemaError $ "List of member types cannot be empty for union type " <> squote name
-  pure $ void $ P.selectionSetUnion name description objs
+  pure $ P.selectionSetUnion name description objs <&>
+    (\objNameAndSelSets ->
+       catMaybes $ flip map objNameAndSelSets $ \(objName, selSet) ->
+        case selSet of
+          [] -> Nothing
+          _  -> Just (G.SelectionInlineFragment $ G.InlineFragment (Just objName) mempty selSet)
+    )
```

Comment:
Could you please add a comment explaining what this code does, and why?

Reply:
Sure, I'll add a comment